registration/include/pcl/registration/gicp_convergence_criteria.h

```diff
@@ -0,0 +1,369 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#pragma once
+
+#include <pcl/registration/convergence_criteria.h>
+#include <pcl/correspondence.h>
+#include <pcl/memory.h>
+#include <pcl/pcl_macros.h>
+
+namespace pcl {
+namespace registration {
+/** \brief @b GICPConvergenceCriteria  represents an instantiation of
+ * ConvergenceCriteria, and implements the following criteria for registration loop
+ * evaluation:
+ *
+ *  * a maximum number of iterations has been reached
+ *  * the transformation (R, t) cannot be further updated (the difference between
+ * current and previous is smaller than a threshold)
+ *  * the Mean Squared Error (MSE) between the current set of correspondences and the
+ * previous one is smaller than some threshold (both relative and absolute tests)
+ *
+ * \note Convergence is considered reached if ANY of the above criteria are met.
+ *
+ * \author Radu B. Rusu
+ * \ingroup registration
+ */
+template <typename Scalar = float>
+class GICPConvergenceCriteria  : public ConvergenceCriteria {
+public:
+  using Ptr = shared_ptr<GICPConvergenceCriteria <Scalar>>;
+  using ConstPtr = shared_ptr<const GICPConvergenceCriteria <Scalar>>;
+
+  using Matrix4 = Eigen::Matrix<Scalar, 4, 4>;
+
+  enum ConvergenceState {
+    CONVERGENCE_CRITERIA_NOT_CONVERGED,
+    CONVERGENCE_CRITERIA_ITERATIONS,
+    CONVERGENCE_CRITERIA_TRANSFORM,
+    CONVERGENCE_CRITERIA_ABS_MSE,
+    CONVERGENCE_CRITERIA_REL_MSE,
+    CONVERGENCE_CRITERIA_NO_CORRESPONDENCES,
+    CONVERGENCE_CRITERIA_FAILURE_AFTER_MAX_ITERATIONS
+  };
+
+  /** \brief Empty constructor.
+   * Sets:
+   *  * the maximum number of iterations to 1000
+   *  * the rotation threshold to 0.256 degrees (0.99999)
+   *  * the translation threshold to 0.0003 meters (3e-4^2)
+   *  * the MSE relative / absolute thresholds to 0.001% and 1e-12
+   *
+   * \param[in] iterations a reference to the number of iterations the loop has ran so
+   * far \param[in] transform a reference to the current transformation obtained by the
+   * transformation evaluation \param[in] correspondences a reference to the current set
+   * of point correspondences between source and target
+   */
+  GICPConvergenceCriteria (const int& iterations,
+                           const Matrix4& transform,
+                           const Matrix4 &previous_transform,
+                           const pcl::Correspondences& correspondences)
+  : iterations_(iterations)
+  , transformation_(transform)
+  , previous_transformation_ (previous_transform)
```

Comment:
Unlike ICP, in GICP `transformation_ ` and `previous_transformation_` are both transform from input cloud. To get the transformation in each iteration, we need both of them.